webhook/resource_url_notifier.go

```diff
@@ -0,0 +1,348 @@
+	}
+
+	return err
+}
+
+func (r *ResourceURLNotifier) Stop(force bool) {
+	r.closed.Break()
+
+	r.mu.Lock()
+	resourceQueues := r.resourceQueues
+	r.resourceQueues = make(map[string]*resourceQueueInfo)
+	r.mu.Unlock()
+
+	for _, rq := range resourceQueues {
+		rq.Stop(force)
+	}
+}
+
+// poster interface
+func (r *ResourceURLNotifier) Process(ctx context.Context, queuedAt time.Time, event *livekit.WebhookEvent) {
+	fields := logFields(event, r.params.URL)
+
+	queueDuration := time.Since(queuedAt)
+	fields = append(fields, "queueDuration", queueDuration)
+
+	if queueDuration > r.params.MaxAge {
+		fields = append(fields, "reason", "age")
+		r.params.Logger.Infow("dropped webhook", fields...)
+
+		if ph := r.getProcessedHook(); ph != nil {
+			whi := webhookInfo(
+				event,
+				queuedAt,
+				queueDuration,
+				time.Time{},
+				0,
+				r.params.URL,
+				true,
+				nil,
+			)
+			if r.params.FieldsHook != nil {
+				r.params.FieldsHook(whi)
+			}
+			ph(ctx, whi)
+		}
+		return
+	}
+
+	sendStart := time.Now()
+	err := r.send(event)
+	sendDuration := time.Since(sendStart)
+	fields = append(fields, "sendDuration", sendDuration)
+	if err != nil {
+		r.params.Logger.Warnw("failed to send webhook", err, fields...)
+	} else {
+		r.params.Logger.Infow("sent webhook", fields...)
+	}
+	if ph := r.getProcessedHook(); ph != nil {
+		whi := webhookInfo(
+			event,
+			queuedAt,
+			queueDuration,
+			sendStart,
+			sendDuration,
+			r.params.URL,
+			false,
+			err,
+		)
+		if r.params.FieldsHook != nil {
+			r.params.FieldsHook(whi)
+		}
+		ph(ctx, whi)
+	}
+}
+
+func (r *ResourceURLNotifier) send(event *livekit.WebhookEvent) error {
+	encoded, err := protojson.Marshal(event)
+	if err != nil {
+		return err
+	}
+	// sign payload
+	sum := sha256.Sum256(encoded)
+	b64 := base64.StdEncoding.EncodeToString(sum[:])
+
+	r.mu.RLock()
+	apiKey := r.params.APIKey
+	apiSecret := r.params.APISecret
+	r.mu.RUnlock()
+
+	at := auth.NewAccessToken(apiKey, apiSecret).
+		SetValidFor(5 * time.Minute).
+		SetSha256(b64)
+	token, err := at.ToJWT()
+	if err != nil {
+		return err
+	}
+	req, err := retryablehttp.NewRequest("POST", r.params.URL, bytes.NewReader(encoded))
+	if err != nil {
+		// ignore and continue
+		return err
+	}
+	req.Header.Set(authHeader, token)
+	// use a custom mime type to ensure signature is checked prior to parsing
+	req.Header.Set("content-type", "application/webhook+json")
+	res, err := r.client.Do(req)
+	if err != nil {
+		return err
+	}
+	_ = res.Body.Close()
+	return nil
+}
+
+func (r *ResourceURLNotifier) sweeper() {
+	ticker := time.NewTicker(r.params.Timeout / 2)
+	defer ticker.Stop()
+
+	for {
+		select {
+		case <-r.closed.Watch():
+			return
+
+		case <-ticker.C:
+			for it := r.resourceQueueTimeoutQueue.IterateAfter(r.params.Timeout); it.Next(); {
+				rqi := it.Item().Value
+				if !rqi.resourceQueue.StopIfIdle(r.params.Timeout) && rqi.tqi != nil {
+					r.resourceQueueTimeoutQueue.Reset(rqi.tqi)
+					continue
+				}
```

Comment:
using `IterateRemoveAfter`atomically marks the expired items deleted to sync the return value from `Reset`

if we got here the items in the queue have all either been delivered or expired so we can unconditionally stop the queue

Reply:
Thank you Paul. Try 3 here in this [commit](https://github.com/livekit/protocol/pull/1000/commits/81bbfcae42288cbdf1cd5844b51c498338ad470b) with feedback incorporated.